jf_agent/main.py

```diff
@@ -498,24 +509,35 @@ def upload_file(filename, path_to_obj, signed_url, local=False):
     for t in threads:
         t.join()
 
+    success = True
     if any(thread_exceptions):
+        # Run through exceptions and inject them into the agent log
+        for exception in thread_exceptions:
+            agent_logging.log_and_print_error_or_warning(
+                logger, logging.ERROR, error_code=0000, msg_args=[exception], exc_info=True
+            )
         print(
             'ERROR: not all files uploaded to S3. Files have been saved locally. Once connectivity issues are resolved, try running the Agent in send_only mode.'
         )
-        return False
+        success = False
 
     # If sending agent config flag is on, upload config.yml to s3 bucket
     if config.send_agent_config:
         config_file_dict = get_signed_url(['config.yml'])['config.yml']
         upload_file('config.yml', config_file_dict['s3_path'], config_file_dict['url'], local=True)
 
-    # creating .done file
-    done_file_path = f'{os.path.join(config.outdir, ".done")}'
-    Path(done_file_path).touch()
-    done_file_dict = get_signed_url(['.done'])['.done']
-    upload_file('.done', done_file_dict['s3_path'], done_file_dict['url'])
+    if success:
+        # creating .done file, only on success
+        done_file_path = f'{os.path.join(config.outdir, ".done")}'
+        Path(done_file_path).touch()
+        done_file_dict = get_signed_url(['.done'])['.done']
+        upload_file('.done', done_file_dict['s3_path'], done_file_dict['url'])
 
-    return True
+    # Upload log files as last step
+    log_file_dict = get_signed_url([agent_logging.LOG_FILE_NAME])[agent_logging.LOG_FILE_NAME]
+    upload_file(agent_logging.LOG_FILE_NAME, log_file_dict['s3_path'], log_file_dict['url'])
+
```

Comment:
Actually I would make the `.done` file be the very last file uploaded, because the existence of that file is what triggers us to begin server-side processing. So if the `.done` file is uploaded before the log file I think it's possible we'll start to process the submission and miss the log file. So I think the last ones should be the log file and then the `.done` file.